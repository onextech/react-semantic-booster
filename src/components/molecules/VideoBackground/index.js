import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import isMobile from 'ismobilejs';
import { MediaCss } from '../../../utils/responsive';


const VideoContainer = styled.div`
  position: absolute;
  width: 100%;
  height: 100%;
  background-attachment: scroll;
  overflow: hidden;
  z-index: -1;
  
  video {
    min-width: 100%;
    min-height: 100%;
    position: relative;
    z-index: 1;
    // poster-related css
    object-fit: cover;
    ${MediaCss.max.sm`
      ${(props) => {
    if (props.poster) {
      return `
            background-image: url(${props.poster.src});
            ${props.poster.bgPos && `background-position: ${props.poster.bgPos}`}
      `;
    }
    return false;
  }}
  `}
  }
  
  .overlay {
    height: 100%;
    width: 100%;
    position: absolute;
    top: 0px;
    left: 0px;
    z-index: 2;
    background: black;
    opacity: 0.5;
  }
  `;

const VideoBackground = ({ video }) => {
  const defaultVideoProps = {
    loop: true,
    muted: true,
  };
  const isDesktop = !isMobile.any;
  if (isDesktop) {
    /**
     * @note
     * To disable video autoplay, autoplay="false" will not work;
     * To remove autoplay the attribute needs to be removed altogether.
     * @link https://developer.mozilla.org/en-US/docs/Web/HTML/Element/video
     * @type {boolean}
     */
    defaultVideoProps.autoPlay = true;
  }

  const { src, poster, ...videoProps } = video;

  const transparentBase64Image = 'data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7';

  return (
    <VideoContainer poster={poster}>
      <div className="overlay" />
      <video
        {...defaultVideoProps}
        {...videoProps}
        poster={transparentBase64Image}>
        <source src={src} type="video/ogg" />
        <source src={src} type="video/webm" />
        <source src={src} type="video/mp4" />
      </video>
    </VideoContainer>
  );
};

VideoBackground.propTypes = {
  video: PropTypes.shape({
    src: PropTypes.string,
    poster: PropTypes.shape({
      src: PropTypes.string.isRequired,
      bgPos: PropTypes.string,
    }).isRequired,
  }).isRequired,
};

export default VideoBackground;
