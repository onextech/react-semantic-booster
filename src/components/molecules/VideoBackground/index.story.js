import React from 'react';
import { storiesOf } from '@storybook/react';
import { Segment, Container } from 'semantic-ui-react';
import VideoBackground from '.';


storiesOf('VideoBackground')
  .add('Default', () => (
    <Segment
      inverted
      vertical
      style={{
        minHeight: 700,
        display: 'flex',
        alignItems: 'center',
        backgroundColor: 'transparent',
      }}>
      <Container>
        <h1>Hello World</h1>
      </Container>
      <VideoBackground
        video={{
          src: 'https://s3-ap-southeast-1.amazonaws.com/storage.onextech.com/web/site/home/24251288.mp4',
          poster: 'https://s3-ap-southeast-1.amazonaws.com/storage.onextech.com/web/site/home/bluenet.jpg',
        }} />
    </Segment>
  ));
