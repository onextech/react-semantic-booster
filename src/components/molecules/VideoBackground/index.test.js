import React from 'react';
import renderer from 'react-test-renderer';
import { render } from 'enzyme';
import VideoBackground from '.';
import { demoVideo } from './index.story';


describe('<VideoBackground>', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(<VideoBackground video={demoVideo} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('has essential classes/elements', () => {
    const wrapper = render(<VideoBackground video={demoVideo} />);
    const overlay = wrapper.find('.overlay');
    const video = wrapper.find('video');
    expect(overlay.length).toBe(1);
    expect(video.length).toBe(1);
  });
});
