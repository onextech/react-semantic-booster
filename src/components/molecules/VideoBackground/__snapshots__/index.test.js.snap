// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<VideoBackground> renders correctly 1`] = `
<div
  className="VideoBackground__VideoContainer-bUrRxM ehWSXR"
  poster={
    Object {
      "bgPos": "140% 50%",
      "src": "https://s3-ap-southeast-1.amazonaws.com/storage.onextech.com/web/site/home/bluenet.jpg",
    }
  }
>
  <div
    className="overlay"
  />
  <video
    autoPlay={true}
    loop={true}
    muted={true}
    poster="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
  >
    <source
      src="https://s3-ap-southeast-1.amazonaws.com/storage.onextech.com/web/site/home/24251288.mp4"
      type="video/ogg"
    />
    <source
      src="https://s3-ap-southeast-1.amazonaws.com/storage.onextech.com/web/site/home/24251288.mp4"
      type="video/webm"
    />
    <source
      src="https://s3-ap-southeast-1.amazonaws.com/storage.onextech.com/web/site/home/24251288.mp4"
      type="video/mp4"
    />
  </video>
</div>
`;
